### PR TITLE
Enabling redirect uri on local host to honor path

### PR DESCRIPTION
**Pull request by trwalke**
Fix for https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/2167

### CODE REVIEW

File: src/client/Microsoft.Identity.Client/Platforms/Features/DefaultOSBrowser/HttpListenerInterceptor.cs [hunk 1]
```diff
@@ -36,12 +38,18 @@ public HttpListenerInterceptor(ICoreLogger logger)
             HttpListener httpListener = null;
             try
             {
-                string urlToListenTo = "http://localhost:" + port + "/";
+                string urlToListenTo = "http://localhost:" + port + (path.StartsWith("/")? path : "/" + path);
+
+                if (!urlToListenTo.EndsWith("/"))
+                {
+                    urlToListenTo += "/";
```
Comment (trwalke) [inline src/client/Microsoft.Identity.Client/Platforms/Features/DefaultOSBrowser/HttpListenerInterceptor.cs:45]: Paths must end with one /

File: src/client/Microsoft.Identity.Client/Platforms/Features/DefaultOSBrowser/HttpListenerInterceptor.cs [hunk 2]
```diff
@@ -36,12 +38,18 @@ public HttpListenerInterceptor(ICoreLogger logger)
             HttpListener httpListener = null;
             try
             {
-                string urlToListenTo = "http://localhost:" + port + "/";
+                string urlToListenTo = "http://localhost:" + port + (path.StartsWith("/")? path : "/" + path);
```
Comment (jennyf19) [inline src/client/Microsoft.Identity.Client/Platforms/Features/DefaultOSBrowser/HttpListenerInterceptor.cs:41]: >path [](start = 69, length = 4)

check for null on `path`

File: tests/Microsoft.Identity.Test.Unit/WebUITests/DefaultOsBrowserWebUiTests.cs [hunk 3]
```diff
@@ -77,7 +76,7 @@ public async Task DefaultOsBrowserWebUi_HappyPath_Async()
             Assert.IsFalse(string.IsNullOrEmpty(authorizationResult.Code));
 
             await _tcpInterceptor.Received(1).ListenToSingleRequestAndRespondAsync(
-                TestPort, Arg.Any<Func<Uri, MessageAndHttpCode>>(), CancellationToken.None).ConfigureAwait(false);
+                TestPort, String.Empty, Arg.Any<Func<Uri, MessageAndHttpCode>>(), CancellationToken.None).ConfigureAwait(false);
```
Comment (jennyf19) [inline tests/Microsoft.Identity.Test.Unit/WebUITests/DefaultOsBrowserWebUiTests.cs:79]: >S [](start = 26, length = 1)

nit: lowercase `s`

File: src/client/Microsoft.Identity.Client/Platforms/Features/DefaultOSBrowser/HttpListenerInterceptor.cs [hunk 4]
```diff
@@ -18,6 +18,7 @@ internal class HttpListenerInterceptor : IUriInterceptor
         public Action TestBeforeTopLevelCall { get; set; }
         public Action TestBeforeStart { get; set; }
         public Action TestBeforeGetContext { get; set; }
+        public Action<string> TestListeningUriBeforeStart { get; set; }
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/Platforms/Features/DefaultOSBrowser/HttpListenerInterceptor.cs:21]: nit: TestBeforeStart could have been modified to emit the url

**Review by bgavrilMS — APPROVED** [review]
Please also manually validate the E2E, i.e. configure an app to use http://localhost/path1/path2 as redirect URI and see if the system browser flow works.

**Review by jennyf19 — APPROVED** [review]
:shipit:

File: tests/Microsoft.Identity.Test.Unit/WebUITests/HttpListenerInterceptorTests.cs [hunk 5]
```diff
@@ -35,18 +36,19 @@ public async Task HttpListenerCompletesAsync()
             // Start the listener in the background
             Task<Uri> listenTask = listenerInterceptor.ListenToSingleRequestAndRespondAsync(
                 port,
+                string.Empty,
```
Comment (neha-bhargava) [inline tests/Microsoft.Identity.Test.Unit/WebUITests/HttpListenerInterceptorTests.cs:39]: nit: lower case s in this class too.

**Review by neha-bhargava — APPROVED** [review]

**Review by henrik-me — APPROVED** [review]
:shipit:

**Pull request merged**